Add capability to do unit conversations to capgen

dustinswales opened the pull request:
This PR adds automatic conversions for supported unit and type transforms. 
Also, included is a new ccpp variable property to indicate a scheme variables vertical orientation vertical, top_at_one. 
As default, top_at_one is set to .false.(GFS ordering convention). 
Adding top_at_one = .true. to a variable in a schemes metadata file will trigger automatic array flipping:

The var_aciton_test was modified to test these new functionalities:
![image](https://github.com/NCAR/ccpp-framework/assets/13949033/e748fb5a-1315-4e1f-bc22-4ba19e04855d)

Addresses https://github.com/NCAR/ccpp-framework/issues/329 and https://github.com/NCAR/ccpp-framework/issues/403

Conversions supported:
https://github.com/NCAR/ccpp-framework/blob/main/scripts/conversion_tools/unit_conversion.py

Commit:
Initial commit. Work in progress

Commit:
Working.

Commit:
Housekeeping

gold2718 commented:
Last week: "meet with us and share how you would have approached it"
This week: PR
Did I miss something?

dustinswales commented:
> Last week: "meet with us and share how you would have approached it" This week: PR Did I miss something?
Fair point. 
This is a draft on how we approached it, which may or may not be inline with your thinking on how to use the VarAction class for unit conversions. Hopefully this makes our upcoming conversation more focused.

climbfuji commented on scripts/metavar.py:
```diff
@@ -1204,6 +1205,47 @@ def add_to_list(self, vlist):
+    def write_action(self, vadict):
+        """Return unit conversion transformation"""
+        if self._set_action:
+            if self._key == 'from':
```
While this works now, a problem with this design comes up with multiple variable actions necessary for one variable. For instance, flipping vertically and doing unit conversions, maybe adding a vloop around it. Trying to come up with a smart code generator that combines all this in one assigment is basically impossible, therefore either one after the other (which then requires other logic so that the rhs in the first assignment is `__local_name`, but all future rhs use `__local_name_temp`, and vice versa for the reverse process after the call to the subroutine), or generate nested function calls that recursively perform these actions. I think the former is easier to implement, but it probably won't allow you to substitute in the variable names in the var action expression until they get written out in lines 1218 / 1223 in scripts/suite_objects.py

dustinswales replied:
Excellent point. I didn't consider multiple actions on the same variable.
The way I do the write step is "by action", w/o consideration for other actions on that variable. But as you say, it should be "by variable", accounting for all actions on the variable.

climbfuji replied:
That's actually the major limitation in `ccpp_prebuild`, I worked around it by coding up all possible combinations (of which there are only four in prebuild, but in capgen there will be a lot more, therefore the need to do better).

dustinswales replied:
Gotcha. I'm starting to understand, albeit late to the party.

climbfuji commented on scripts/metavar.py:
```diff
@@ -1204,6 +1205,47 @@ def add_to_list(self, vlist):
+
+        return f"{lhs} = {rhs}"
+
+    @property
```
Do you need lines 1237-1245? I don't see where this gets used.

climbfuji commented on scripts/metavar.py:
```diff
@@ -1823,6 +1865,32 @@ def __eq__(self, other):
+            ucmatch["from"] = VarUnitConv(function_from(),key="from",
+                                          local_name=var.get_prop_value('local_name'),
+                                          kind=var.get_prop_value('kind'))
+        except:
```
I think we want to generate an error and abort if a requested unit conversion is not implemented.

dustinswales replied:
Agreed. I was waiting until things were more settled before adding in error checking

climbfuji commented on scripts/metavar.py:
```diff
@@ -1823,6 +1865,32 @@ def __eq__(self, other):
+                                            local_name=var.get_prop_value('local_name'),
+                                            kind=var.get_prop_value('kind'))
+        except:
+            ucmatch["to"]   = None
```
Ditto: I think we want to generate an error and abort if a requested unit conversion is not implemented.

dustinswales commented on scripts/metavar.py:
```diff
@@ -1204,6 +1205,47 @@ def add_to_list(self, vlist):
+            setattr(self, "_"+key, value)
+        self._set_action = set_action
+
+    def add_local(self, var):
```
Yeah this is not used. I will remove

dustinswales commented on scripts/metavar.py:
```diff
@@ -1184,7 +1185,7 @@ def equiv(self, vmatch):
         equiv should be overridden with a method that first calls this
         method and then tests class-specific object data."""
-        return vmatch.__class__ == self.__class__
+        return vmatch.__class__ != self.__class__
```
This was a (bad) hack for _add_to_list to work. 
Now that I'm thinking about this considering "multiple actions for each variable", there will be some more changes coming. This will go away

Commit:
Working with compatability object

Commit:
Merge branch 'feature_capgen_unit_conversions_temp' into feature_capgen_unit_conversions

Commit:
Revert change

Commit:
Omission from previous commit

Commit:
Omission from previous commit. again...

Commit:
Omission from previous commit. again...

Commit:
Merge branch 'feature/capgen' of https://github.com/NCAR/ccpp-framework into feature_capgen_unit_conversions

dustinswales commented:
@DomHeinzeller @gold2718 @peverwhee 
As we discussed, I jettisoned what I was doing w/ the VarAction class and added the unit-conversions to the compatibility object.
This works for unit conversions, but may need some changes to flip the vertical orientation. 
A related question, how do we recognize when a variable needs to be flipped in the vertical? (New attributes in host and suite for convention?)

gold2718 commented:
> @DomHeinzeller @gold2718 @peverwhee As we discussed, I jettisoned what I was doing w/ the VarAction class and added the unit-conversions to the compatibility object.

It looks like you added a `VarCompatObject` to what the unit conversion work. Is that what you meant?

> This works for unit conversions, but may need some changes to flip the vertical orientation. A related question, how do we recognize when a variable needs to be flipped in the vertical? (New attributes in host and suite for convention?)

This has been recognized for quite a while. I know there was a discussion in a Google doc many years ago but I no longer have access to that (if I could even remember where it was).

However @grantfirl started this back up in August and I would like to see some progress on this so please refer to the [Discussion related to vertical coordinate interoperabilty](https://github.com/NCAR/ccpp-framework/discussions/487). I have added a bit of history to that.

dustinswales committed:
Cleanup VarCompatObj creation

grantfirl committed:
Bug fix for initialized_set_block (#503)

dustinswales committed:
Vertical flipping working w/ new metadata attribute. Work in progress

dustinswales committed:
Some more changes. Vertical flipping working

dustinswales committed:
Housekeeping

dustinswales committed:
Add kind transform to transform test

dustinswales committed:
Add local declaration of vertical extent, needed for vertical flipping

dustinswales commented:
Hey @gold2718 @DomHeinzeller @peverwhee 
I think this is at a point where I'd like to get some feedback from the team. 
- Unit, kind, and vertical loop ordering conversions are working using VarCompatObj.
- There's a new variable attribute, top_at_one, which is used to trigger vertical loop flipping.
- I modified the test/var_action_test to trigger these three types of conversions:
![Screenshot 2023-10-27 at 11 43 00 AM](https://github.com/NCAR/ccpp-framework/assets/13949033/df845a49-8266-4eed-bc11-45894020ebb2)

Questions/Comments/ToDo:
- I kinda obliterated the forward/reverse_transform functions in VarCompatObj to make this work. I don't the idea behind why they were designed the way they were(My guess is in the context of a vertical loop substitution?). Also, they weren't used anywhere other in doctests, so I felt that messing with them was okay.
- Modify test to have vertically dependent fields (to test vertical flipping). This is trivial.
- dimension reordering action (e.g. [i,j] -> [j,i])? This is built into varCompatObj, but so were the other transforms, which needed refactoring to work. Is this a priority?
- This is not as clean as I'd like and I'm not sure how to best organize this? Specifically, the declarations and allocations of local variables occur at the Group level, whereas the actions are written at the Scheme level. I think it would be cleaner if all transform related business happened at the same place. Maybe this is not feasible? (I haven't looked hard into this)

gold2718 commented on scripts/suite_objects.py:
```diff
@@ -1184,6 +1194,41 @@ def analyze(self, phase, group, scheme_library, suite_vars, level):
+                    for dpart in dim[0].split(':'):
+                        indices[dim[1]] = var.get_prop_value('local_name')+'_nlay:1:-1'
+                    # Create statement for use in write stage.
+                    write_stmt = var.get_prop_value('local_name')+"_nlay = size({},{})"
```
I don't think this is valid. I do not think there is any requirement that the array size has to match the dimension. With Fortran 90-style array declarations (as in the var action test), you should be able to get correct behavior even if they do not match. Also, I do not see why you need a separate local variable. Why just not use `nlev`?

gold2718 reviewed (commented):
See comments / questions below.
Also, note that the current var_action unit test suite does not all pass. This suite needs a careful look to make sure it is still correct and also should be expanded for more complex tests.

dustinswales committed:
Everything but getting the correct local variable name for the vertical dimension name

climbfuji reviewed (commented):
I realized that I forgot to hit the "submit review" button. Most of it is outdated by now, but for the records ...

dustinswales commented on scripts/suite_objects.py:
```diff
@@ -1184,6 +1192,31 @@ def analyze(self, phase, group, scheme_library, suite_vars, level):
+                indices = [':']*var.get_rank()
+                if compat_obj.has_vert_transforms:
+                    dim = find_vertical_dimension(var.get_dimensions())
+                    lvar = 'nlev'
```
@gold2718 @climbfuji @peverwhee 
How do we populate "lvar" with the correct local_name, "nlev", for the standard_name "vertical_layer_dimension"?
(Please someone help me out with the magic needed to do this!!!!)

As far as I can tell, we are in the variable loop within analyze, and we don't have access to the other variables(?), just the current one, so my workaround was to create a local variable using the current variable in the loop, local_array_nlay = size(local_array,dim). But this is not ideal from our conversations (I'm still not convinced why, but anyways...)

gold2718 replied:
Sorry for the delay, busy day . . .
Inside the block `if compat_obj is not None and (compat_obj.has_vert_transforms or compat_obj.has_unit_transforms):', I tried:
```
                vdim_name = vert_dim.split(':')[-1]
                group_vvar = self.__group.call_list.find_variable(vdim_name)
                vname = vvar.get_prop_value('local_name')
```
which gets me `nlev`. Is this what you were looking for?

dustinswales replied:
No worries at all. That piece did the trick, nice and slick. Thanks!
![Screenshot 2023-10-31 at 1 48 45 PM](https://github.com/NCAR/ccpp-framework/assets/13949033/c16f9c59-b20a-4d81-9d89-2c5d725a29e0)

dustinswales commented on scripts/suite_objects.py:
```diff
@@ -1123,13 +1130,16 @@ def analyze(self, phase, group, scheme_library, suite_vars, level):
-            args = self.match_variable(var, vstdname=vstdname, vdims=vdims)
-            found, vert_dim, new_dims, missing_vert = args
+            args = self.match_variable(var, self.run_env)
+            found, vert_dim, new_dims, missing_vert, compat_obj, suite_var = args
```
It's not, but maybe it is needed to do what I'm not doing?

dustinswales committed:
Add snippet from Steve G. to get correct local_name for vertical_layer_dimension.

Commit:
Address reviewers comments

Commit:
Address reviewers comments again

Commit:
Address reviewers comments again again

Commit:
Merge branch 'feature/capgen' of https://github.com/NCAR/ccpp-framework into feature_capgen_unit_conversions

climbfuji commented:
@dustinswales Can you please merge in feature/capgen (need to resolve conflicts) before I review? Thanks!

climbfuji committed:
Rename tests directory to test_prebuild

climbfuji committed:
Add test for ccpp_prebuild for blocked data structures

climbfuji committed:
Add CI tests for ccpp-prebuild

climbfuji committed:
Debugging CI workflow for prebuild

climbfuji committed:
Self-review

Commit:
Merge branch 'main' of https://github.com/NCAR/ccpp-framework into feature_capgen_unit_conversions

Commit:
Merge branch 'feature/capgen' of https://github.com/NCAR/ccpp-framework into feature_capgen_unit_conversions

Commit:
Revert "Merge branch 'main' of https://github.com/NCAR/ccpp-framework into feature_capgen_unit_conversions"

This reverts commit 998cc7bf97ac39d5682092f6fd14ca94b09adb3e, reversing
changes made to 3b045946dcf6bdfe70afa9603233eef2ab3b25cf.

Commit:
Activate var_action_test

climbfuji commented on scripts/suite_objects.py:
```diff
@@ -1598,7 +1639,7 @@ def register_action(self, vaction):
             vaction.add_local(self, _API_LOCAL, self.run_env)
             return True
-        # end if
+
```
I think we want the original `# end if` here as per our conversation last week

dustinswales replied:
done.

climbfuji commented on scripts/var_props.py:
```diff
@@ -926,6 +927,13 @@ def __init__(self, var1_stdname, var1_type, var1_kind, var1_units,
         # end if
+        if self.__compat:
+            # Check for vertical array flipping (do later)
+            if var1_top or var2_top:
```
Don't we have to exlude `var1_top and var2_top`?

dustinswales replied:
Not sure I follow?
The assumed convention is that top_at_one=False. So any variable (host or scheme) that has top_at_one=True will get transformed before/after calling the schme.

climbfuji replied:
Maybe I am confused. if `top_at_one=True` for both the host model variable and the scheme variable, then there's no need to transform?

dustinswales replied:
Ah. I see your point. Good catch.
Fixed.

climbfuji commented on test/var_action_test/effr_calc.F90:
```diff
@@ -23,7 +23,7 @@ subroutine effr_calc_run(ncol, nlev, effrr_in, effrl_inout, &
       real(kind_phys),    intent(inout) :: effrl_inout(:,:)
       real(kind_phys),    intent(out)   :: effri_out(:,:)
-      real(kind_phys),    intent(inout) :: effrs_inout(:,:)
+      real(8),intent(inout) :: effrs_inout(:,:)
```
Why? I think this should be reverted. Or is this something you want to test for explicitly (now seeing the change in the meta file)?

dustinswales replied:
This is needed to test the kind transformations.

climbfuji replied:
Ok, thanks!

climbfuji reviewed (changes requested):
Good stuff!

climbfuji commented:
In general, it would be nice to have more inline documentation in form of comments and/or docstrings, especially for newly added code. `capgen` is a pretty complicated piece of python code, any documentation will be greatly appreciated.

climbfuji committed:
test_prebuild/test_blocked_data: 'use, intrinsic :: iso_fortran_env, only: error_unit'

Commit:
Fix unit tests for variable transformations. Reorganization a bit to be consistent with how the DEBUG checking is implemented in PR #512

Commit:
Cleanup forward/reverse transform terminology

Commit:
Doc tests for transformations

Commit:
Add more inline documentation. Address reviewers comments.

dustinswales commented on scripts/suite_objects.py:
```diff
@@ -1139,13 +1144,16 @@ def analyze(self, phase, group, scheme_library, suite_vars, level):
         # end if
         scheme_mods = set()
         scheme_mods.add((my_header.module, self.subroutine_name))
+        var_local = {"local_name":[], "std_name":[]}
```
This was a relic. Removed.

dustinswales commented on scripts/suite_objects.py:
```diff
@@ -1223,13 +1258,19 @@ def write(self, outfile, errcode, indent):
     def schemes(self):
         """Return self as a list for consistency with subcycle"""
         return [self]
+        # end if
```
done

dustinswales commented:
@DomHeinzeller I added some more inline documentation and addressed your comments. Since you reviewed, I refactored a bit to be more closely aligned with #512.
@gold2718 @peverwhee @mkavulich This is ready for review.

Commit:
Address reviewer comment

climbfuji commented on scripts/var_props.py:
```diff
@@ -929,7 +921,7 @@ def __init__(self, var1_stdname, var1_type, var1_kind, var1_units,
         if self.__compat:
             # Check for vertical array flipping (do later)
-            if var1_top or var2_top:
+            if (var1_top or var2_top) and (var1_top != var2_top):
```
```suggestion
            if var1_top != var2_top:
```
should do

dustinswales replied:
Done

Commit:
Update authors in CMakeList.

Commit:
Update authors in CMakeList.

Commit:
Address reviewer comment

climbfuji committed:
Merge pull request #515 from climbfuji/feature/ccpp_prebuild_blocked_data_ci_tests

ccpp-prebuild: add blocked data tests and run prebuild tests in CI

peverwhee commented on scripts/var_props.py:
```diff
@@ -1007,16 +1003,10 @@ def reverse_transform(self, lvar_lname, rvar_lname, indices,
-            lhs_term = f"{lvar_lname}({','.join(indices)})"
-        # end if
-        rhs_term = f"{rvar_lname}({','.join(indices)})"
+        # Dimension transforms (Indices handled exrernally)
```
typo - externally

dustinswales replied:
done.

peverwhee commented on scripts/var_props.py:
```diff
@@ -945,13 +945,15 @@ def __init__(self, var1_stdname, var1_type, var1_kind, var1_units,
-           will show up in the RHS of the transform as "var1(indices)".
+        <rvar_indices> is a tuple of the loop indices for "var1" (i.e., "var1"
+           will show up in the RHS of the transform as "var1(rvar_indices)".
+        <lvar_indices> is a tuple of the loop indices for "var1" (i.e., "var2"
```
var2?

dustinswales replied:
done

peverwhee commented on scripts/var_props.py:
```diff
@@ -989,13 +985,13 @@ def forward_transform(self, lvar_lname, rvar_lname, indices,
-        <indices> is a tuple of the loop indices for "var2" (i.e., "var2"
-           will show up in the RHS of the transform as "var2(indices)".
+        <rvar_indices> is a tuple of the loop indices for "var1" (i.e., "var1"
+           will show up in the RHS of the transform as "var1(rvar_indices)".
```
Should add documentation for <lvar_indices> here

dustinswales replied:
done

peverwhee commented on scripts/suite_objects.py:
```diff
@@ -1200,6 +1205,12 @@ def analyze(self, phase, group, scheme_library, suite_vars, level):
                                                   vstdname))
                 # end if
             # end if
+            # Are there any forward/reverse transforms for this variable?
+            if compat_obj is not None and (compat_obj.has_vert_transforms or
+                                           compat_obj.has_unit_transforms or
+                                           compat_obj.has_kind_transforms):
+                self.add_var_transform(var, compat_obj, vert_dim)
```
do we need to do something if the variables aren't compatible at all? I see that we're logging the reason for incompatibility in compatible() but I don't see an endrun or anything like that.

dustinswales replied:
It happens much earlier, during __init in VarCompatObj, if there are no valid conversions it will report errors then. For example,
-- Unsupported unit conversion, 'umsdsxc' to 'm' for 'effective_radius_of_stratiform_cloud_ice_particle'
-- Unknown DDT type, real_crazy, at /glade/u/home/dswales/Projects/framework/capgen/unit_conversions/ccpp-framework/test/var_action_test/effr_calc.meta:42

peverwhee replied:
Ah yes, thanks Dustin!

Follow-up question, how does it work for what the code calls "fatal incompatibilities" like standard name and type? It looks like, for those, the incompatibility reasons are logged but it the logic (in the init phase of VarCompatObj) precludes an error from being thrown (errors are thrown by _get_kind_convstrs).

dustinswales replied:
The fatal incompatibilities are caught before we try to do a transform, if "not found".

peverwhee replied:
ok. Thanks again, Dustin!

peverwhee reviewed (changes requested):
Thanks Dustin! A couple things.

Also: would it make sense to rename the var_action_test to something else? var_compatibility_test?

Commit:
Address comments

Commit:
Rename var_action test var_compatability

dustinswales commented:
@peverwhee Comments addressed. Thanks!
I went ahead and renamed the test var_compatability.

Commit:
Merge branch 'main' of https://github.com/NCAR/ccpp-framework into feature_capgen_unit_conversions

peverwhee reviewed (approved):
looks good!

climbfuji commented:
@mwaxmonsky @gold2718 When you have time, can you please review this PR? Thanks very much in advance.

climbfuji commented:
Pinging all remaining reviewers. This PR has sufficient approvals and we are planning to merge this PR on Friday, January 12 2024.
